certbot/lock.py

```diff
@@ -0,0 +1,139 @@
+"""Implements file locks for locking files and directories in UNIX."""
+import errno
+import fcntl
+import logging
+import os
+
+from certbot import errors
+
+logger = logging.getLogger(__name__)
+
+
+def lock_dir(dir_path):
+    """Place a lock file on the directory at dir_path.
+
+    The lock file is placed in the root of dir_path with the name
+    .certbot.lock.
+
+    :param str dir_path: path to directory
+
+    :returns: the locked LockFile object
+    :rtype: LockFile
+
+    :raises errors.LockError: if unable to acquire the lock
+
+    """
+    return LockFile(os.path.join(dir_path, '.certbot.lock'))
+
+
+class LockFile(object):
+    """A UNIX lock file.
+
+    This lock file is released when the locked file is closed or the
+    process exits. It cannot be used to provide synchronization between
+    threads. It is based on the lock_file package by Martin Horcicka.
+
+    """
+    def __init__(self, path):
+        """Initialize and acquire the lock file.
+
+        :param str path: path to the file to lock
+
+        :raises errors.LockError: if unable to acquire the lock
+
+        """
+        super(LockFile, self).__init__()
+        self._path = path
+        self._fd = None
+
+        self.acquire()
+
+    def acquire(self):
+        """Acquire the lock file.
+
+        :raises errors.LockError: if lock is already held
+        :raises OSError: if unable to open or stat the lock file
+
+        """
+        while self._fd is None:
+            # Open the file
+            fd = os.open(self._path, os.O_CREAT | os.O_WRONLY, 0o600)
+            try:
+                self._try_lock(fd)
+                if self._lock_success(fd):
+                    self._fd = fd
+            finally:
+                # Close the file if it is not the required one
+                if self._fd is None:
+                    os.close(fd)
+
+    def _try_lock(self, fd):
+        """Try to acquire the lock file without blocking.
+
+        :param int fd: file descriptor of the opened file to lock
+
+        """
+        try:
+            fcntl.lockf(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+        except IOError as err:
+            if err.errno in (errno.EACCES, errno.EAGAIN):
+                logger.debug(
+                    "A lock on %s is held by another process.", self._path)
+                raise errors.LockError(
```

Comment:
Wait, this errors out but it's in a while loop. So if we fail to acquire, we don't just quit, we try again to grab the lock until we manage to successfully grab it. Yes?